ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/ValidateDisclosedContracts.scala

```diff
@@ -0,0 +1,114 @@
+import com.daml.ledger.api.validation.ValidationErrors.invalidArgument
+import com.daml.ledger.api.validation.ValueValidator.{
+  validateOptionalIdentifier,
+  validateRecordFields,
+}
+import com.daml.lf.value.Value.ValueRecord
+import com.daml.platform.server.api.validation.FieldValidations.{
+  requireContractId,
+  requirePresence,
+  validateHash,
+  validateIdentifier,
+}
+import com.google.protobuf.timestamp.Timestamp
+import io.grpc.StatusRuntimeException
+import com.daml.lf.data.Time
+
+import scala.collection.mutable
+import scala.util.Try
+
+class ValidateDisclosedContracts(featureEnabled: Boolean) {
+  def apply(commands: ProtoCommands)(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, ImmArray[DisclosedContract]] =
+    for {
+      _ <- Either.cond(
+        featureEnabled || commands.disclosedContracts.isEmpty,
+        (),
+        LedgerApiErrors.RequestValidation.InvalidField
+          .Reject(
+            "disclosed_contracts",
+            "feature in development: disclosed_contracts should not be set",
+          )
+          .asGrpcError,
+      )
+      validatedDisclosedContracts <- validateDisclosedContracts(commands.disclosedContracts)
+    } yield validatedDisclosedContracts
+
+  private def validateDisclosedContracts(
+      disclosedContracts: Seq[ProtoDisclosedContract]
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, ImmArray[DisclosedContract]] = {
+    type ZeroType =
+      Either[
+        StatusRuntimeException,
+        mutable.Builder[DisclosedContract, ImmArray[DisclosedContract]],
+      ]
+
+    disclosedContracts
+      .foldLeft[ZeroType](Right(ImmArray.newBuilder))((contractz, contract) =>
+        for {
+          validatedContracts <- contractz
+          validatedContract <- validateDisclosedContract(contract)
+        } yield validatedContracts.addOne(validatedContract)
+      )
+      .map(_.result())
+  }
+
+  private def validateDisclosedContract(
+      disclosedContract: ProtoDisclosedContract
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[StatusRuntimeException, DisclosedContract] =
+    for {
+      templateId <- requirePresence(disclosedContract.templateId, "template_id")
+      validatedTemplateId <- validateIdentifier(templateId)
+      contractId <- requireContractId(disclosedContract.contractId, "contract_id")
+      createArguments <- requirePresence(disclosedContract.arguments, "arguments")
+      recordId <- validateOptionalIdentifier(createArguments.recordId)
+      validatedRecordField <- validateRecordFields(createArguments.fields)
+      metadata <- requirePresence(disclosedContract.metadata, "metadata")
+      createdAt <- requirePresence(metadata.createdAt, "created_at")
+      validatedCreatedAt <- validateCreatedAt(createdAt)
+      keyHash <- validateHash(metadata.contractKeyHash, "contract_key_hash")
```

Comment:
what about the required presence of the driver_metadata?

Reply:
Specified as `driver_metadata` as `Optional` in the proto docs: 245a9b0c60d2d76d628f0b53c828bdbbfc7d521e

Reply:
Since an empty byte array is the same as a missing `driver_metadata` field and we will encounter empty driver metadata in practice, I would say it must be optional.